src/cli.c

```diff
@@ -148,22 +149,70 @@ static void send_header(quicly_stream_t *stream, int is_http1, int status, const
+    int fd;
+    struct stat st;
 
-    if ((fp = fopen(fn, "rb")) == NULL)
+    if ((fd = open(fn, O_RDONLY)) == -1)
         return 0;
-    send_header(stream, is_http1, 200, mime_type);
-    while ((n = fread(buf, 1, sizeof(buf), fp)) != 0)
-        quicly_streambuf_egress_write(stream, buf, n);
-    fclose(fp);
+    if (fstat(fd, &st) != 0) {
+        close(fd);
+        return 0;
+    }
 
+    send_header(stream, is_http1, 200, mime_type);
+    quicly_sendbuf_vec_t vec = {&send_file_callbacks, (size_t)st.st_size, (void *)(intptr_t)fd};
+    quicly_streambuf_egress_write_vec(stream, &vec);
     return 1;
 }
 
+static int flatten_sized_text(quicly_sendbuf_vec_t *vec, void *dst, size_t off, size_t len)
+{
+    static const char pattern[] =
+        "hello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\n";
+
+    assert(len < sizeof(pattern) - 13); /* pattern is bigger than MTU size */
+    memcpy(dst, pattern + off % 12, len);
```

Comment:
Hm, not sure I follow. I understand that you want to copy from the beginning of the repeating pattern, but when does the caller ask for a non-zero offset? The API of this function is unclear to me.

Reply:
This is an implementation of the quicly_sendbuf_flatten_vec_cb callback. Would adding a reference to the callback type (that has the doc-comments explaining the interface) resolve the concern?

Reply:
Done in b8ba132.